rfc/text/advanced/ap_35_authentication_cryptosign.md

```diff
@@ -1,3 +1,1124 @@
+
+and MAY implement one or more of
+
+* [TLS Channel Binding](#channelbinding)
+* [Router Authentication](#routerauth)
+* [Trustroots and Certificates](#trustrootcerts)
+* [Remote Attestation](#remoteattestation)
+
+The following sections describe each of above features of *WAMP-Cryptosign*.
+
+Examples of complete authentication message exchanges can be found at the end of this
+chapter in [Example Message Exchanges](#examplemessageexchanges).
+
+
+### Client Authentication {#clientauth}
+
+A *Client* is authenticated to a *Router* by:
+
+1. sending a `HELLO`, announcing its public key
+2. signing a (random) challenge received in `CHALLENGE` with its private key
+3. let the router verify the signature, proofing the client actually controls the private key, and thus the authenticity of the client as identified by the public key
+4. let the router admit the client to a realm under a role, based on the public key
+
+Thus, the client is identified using its public key, and the *Router* needs to know said public key and its desired realm and role in advance.
+
+A *Client* for which the *Router* does not previously know the client's public key MAY use the [Trustroots and Certificates](#name-trustroots-and-certificates) feature to trust a *Client* based on an additional certificate presented by the client.
+
+A *Router* is optionally (see [Router Authentication](#name-router-authentication)) authenticated to a *Client* by:
+
+1. client includes a (random) `HELLO.Details.challenge|string`
+2. the router sends the signature as part of its challenge to the client, in `CHALLENGE.extra.signature|string`
+
+Again, in this case, the *Router* includes a trustroot and certificate for the client to verify.
+
+
+#### Computing the Signature
+
+The challenge sent by the router is a 32 bytes random value, encoded as a Hex string in `CHALLENGE.extra.challenge|string`.
+
+When no channel binding is active, the Ed25519 signature over the 32 bytes message MUST be computed using the WAMP-Cryptosign *private key* of the authenticating client.
+
+When channel binding is active, the challenge MUST first be XOR'ed bytewise with the channel ID, e.g. the 32 bytes from TLS with channel binding `"tls-unique"`, and the resulting message (which again has length 32 bytes) MUST be signed using the WAMP-Cryptosign *private key* of the authenticating client.
+
+The client MUST return the concatenation of the signature and the message signed (96 bytes) in the `AUTHENTICATE` message.
+
+
+#### Example Message Flow
+
+A typical authentication begins with the client sending a `HELLO` message specifying the `cryptosign` method as (one of) the authentication methods:
+
+{align="left"}
+```javascript
+[1, "realm1", {
+    "roles": {/* see below */},
+    "authmethods": ["cryptosign"],
+    "authid": "client01@example.com",
+    "authextra": {
+        "pubkey": "545efb0a2192db8d43f118e9bf9aee081466e1ef36c708b96ee6f62dddad9122",
+        "channel_binding": null
```

Comment:
As `channel_binding` is optional and is not filled in this example, I suggest omitting it completely for the clearance.